src/django_components/html.py

```diff
@@ -1,101 +1,100 @@
-
-
-# TODO: Remove `transform_html_document`, `insert_before_end`, `html_insert_before_end`
-#       Once Selectolax supports `LaxborNode.insert_child()`
-#       See https://github.com/rushter/selectolax/issues/128
-def transform_html_document(html: str, transform: TransformHtmlCallback) -> str:
-    # Escape <!doctype>, <html>, <head>, and <body> tags, because Selectolax treats them
-    # specially, which makes it impossible to edit them.
-    def on_replace_match(match: "re.Match[str]") -> str:
-        match_str = match[0].lower()
-        return html_tag_escapes[match_str]
-
-    escaped_html = HTML_ROOT_TAGS_REGEX.sub(on_replace_match, html)
-
-    # Selectolax now treats the escaped tags as custom tags, and so Selectolax wraps
-    # the whole content in extra <html><head></head><body> CONTENT </body></html>.
-    # So the actual HTML is under `.body.child`
-    wrapper_tree = LexborHTMLParser(escaped_html).body
-    escaped_tree = wrapper_tree.child if wrapper_tree else None
-    if escaped_tree:
-        body_matches = escaped_tree.css("body_")
-        head_matches = escaped_tree.css("head_")
-        body = body_matches[0] if body_matches else None
-        head = head_matches[0] if head_matches else None
+    if is_fragment:
+        _, nodes = parse_multiroot_html(html)
+        return nodes
     else:
-        body = None
-        head = None
-
-    # Finally, we can pass the actual <head> and <body> tags to be transformed
-    # The transformations happen in-place in Selectolax.
-    transform(head, body)
-    transformed_html = cast(str, escaped_tree.html) if escaped_tree else ""
-
-    # After the transformations are applied, we need to un-escape the <doctype_>, <head_>, etc tags
-    def on_reverse_replace_match(match: "re.Match[str]") -> str:
-        match_str = match[0].lower()
-        return html_tag_escapes_reverse[match_str]
-
-    final_html = HTML_ROOT_TAGS_REVERSE_REGEX.sub(on_reverse_replace_match, transformed_html)
-
-    # NOTE: Because of how we work around selectolax to enable modifying the <body> and <head>
-    # tags, selectolax also adds `</doctype_>` at the end of the content when we serialize the HTML.
-    # So we have to remove that.
-    return final_html[: -len("</doctype_>")]
+        return tree
+
+
+def parse_multiroot_html(html: str) -> Tuple[LexborNode, List[LexborNode]]:
+    """
+    Use this when you know the given HTML is a multiple nodes like
+
+    `<div> Hi </div> <span> Hello </span>`
+    """
+    # NOTE: HTML / XML MUST have a single root. So, to support multiple
+    # top-level elements, we wrap them in a dummy singular root.
+    parser = LexborHTMLParser(f"<root>{html}</root>")
+
+    # Get all contents of the root
+    root_elem = parser.css_first("root")
+    elems = [*root_elem.iter()] if root_elem else []
+    return cast(LexborNode, root_elem), elems
+
+
+def is_html_parser_fragment(html: str, tree: LexborHTMLParser) -> bool:
```

Comment:
One thing that could still be improved in selectolax is that if we use the `HTMLParser`, then it expects a well-formatted HTML with `<html>`, <head>`, and `<body>` tags. And it adds them if these are missing.

So then I have to do the workaround below to figure out which one we're dealing with.

I've raised this https://github.com/rushter/selectolax/issues/74#issuecomment-2404470344